src/protocol/malicious.rs

```diff
@@ -0,0 +1,306 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::{fabric::Network, Direction},
+    protocol::{check_zero::check_zero, context::ProtocolContext, reveal::reveal, RecordId},
+    secret_sharing::Replicated,
+};
+use futures::future::try_join;
+
+use serde::{Deserialize, Serialize};
+
+/// A message sent by each helper when they've computed one share of the result
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct UValue<F> {
+    payload: F,
+}
+
+#[allow(dead_code)]
+pub struct SecurityValidator<'a, F, N> {
+    internally_indexed_ctx: ProtocolContext<'a, N>,
+    r_share: Replicated<F>,
+    u: F,
+    w: F,
+}
+
+impl<'a, F: Field, N: Network> SecurityValidator<'a, F, N> {
+    #[allow(dead_code)]
+    pub fn new(ctx: &ProtocolContext<'a, N>) -> SecurityValidator<'a, F, N> {
+        let internally_indexed_ctx = ctx.narrow(&"ValidationInternalIndexed".to_string());
+        let prss = internally_indexed_ctx.prss();
+
+        let r_share = prss.generate_replicated(RecordId::from(0));
+        let (u_left, u_right): (F, F) = prss.generate_fields(RecordId::from(1));
+        let (w_left, w_right): (F, F) = prss.generate_fields(RecordId::from(2));
+
+        SecurityValidator {
+            internally_indexed_ctx,
+            r_share,
+            u: u_right - u_left,
+            w: w_right - w_left,
+        }
+    }
+
+    #[allow(dead_code)]
+    fn compute_dot_product_contribution(a: Replicated<F>, b: Replicated<F>) -> F {
+        let (a_left, a_right) = a.as_tuple();
+        let (b_left, b_right) = b.as_tuple();
+        (a_left + a_right) * (b_left + b_right) - a_right * b_right
+    }
+
+    #[allow(dead_code)]
+    fn accumulate_macs(
+        &mut self,
+        ctx: &ProtocolContext<'_, N>,
+        record_id: RecordId,
+        x: Replicated<F>,
+        rx: Replicated<F>,
+    ) {
+        // The helpers need to use the same shared randomness to generate the random constant used to validate a given multiplication.
+        // This is a bit tricky, because we cannot count on the multiplications being executed in the same order across all the helpers.
+        // The easiest way is to just narrow the context used to perform the multiplication, and then re-use the same record_id.
+        // This ensures that when the helpers all go to validate the multiplication: "1/foo/bar/baz", they all use the prss from "1/foo/bar/baz/validate".
+        // That way, we don't need to worry about the order in which the multiplications are executed.
+        let prss = ctx.narrow(&"validate".to_string()).prss();
+        let random_constant = prss.generate_replicated(record_id);
+
+        self.u += Self::compute_dot_product_contribution(random_constant, rx);
```

Comment:
As we discussed today, probably the best answer is to merge these two and use streams for this.

Reply:
...after some discussion with @martinthomson - we might not need streams at all. It could be that Arc<Mutex<>> is just as performant and less complex. I might wind up eliminating the other dot-product protocol entirely (but I would do that in another diff)